traffic_router/connector/src/main/java/com/comcast/cdn/traffic_control/traffic_router/secure/CertificateDataConverter.java

```diff
@@ -49,6 +81,76 @@ public HandshakeData toHandshakeData(final CertificateData certificateData) {
 		return null;
 	}
 
+	public boolean verifySubject(final X509Certificate certificate, final String hostAlias ) {
+		final String host = certificate.getSubjectDN().getName();
+		if (hostCompare(hostAlias,host)) {
+			return true;
+		}
+
+		try {
+			// This approach is probably the only one that is JDK independent
+			if (certificate.getSubjectAlternativeNames() != null) {
+				for (final List<?> altName : certificate.getSubjectAlternativeNames()) {
+					if (hostCompare(hostAlias, (String) altName.get(1))) {
+						return true;
+					}
+				}
+			}
+		}
+		catch (Exception e) {
+			log.error("Encountered an error while validating the certificate subject for service: "+hostAlias+", " +
+					"error: "+e.getClass().getSimpleName()+": " + e.getMessage(), e);
+			return false;
+		}
+
+		return false;
+	}
+
+	private boolean hostCompare(final String hostAlias, final String subject) {
+		if (hostAlias.contains(subject) || subject.contains(hostAlias)) {
+			return true;
+		}
+		final String[] chopped = subject.split("CN=", 2);
+		if (chopped != null && chopped.length > 1) {
+			String chop = chopped[1];
+			chop = chop.replaceFirst("\\*\\.", ".");
+			chop = chop.split(",", 2)[0];
+			if (chop.length()>0 && (hostAlias.contains(chop) || chop.contains(hostAlias))) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public boolean verifyModulus(final PrivateKey privateKey, final X509Certificate certificate) {
+		BigInteger privModulus = null;
+		if (privateKey instanceof BCRSAPrivateCrtKey) {
+			privModulus = ((BCRSAPrivateCrtKey) privateKey).getModulus();
+		} else if (privateKey instanceof RSAPrivateCrtKeyImpl) {
+			privModulus = ((RSAPrivateCrtKeyImpl) privateKey).getModulus();
+		}
+		BigInteger pubModulus = null;
+		final PublicKey publicKey = certificate.getPublicKey();
+		if ((publicKey instanceof RSAPublicKeyImpl)) {
+			pubModulus = ((RSAPublicKeyImpl) publicKey).getModulus();
+		} else {
+			final String[] keyparts = publicKey.toString().split(System.getProperty("line.separator"));
+			for (final String part : keyparts) {
+				final int start = part.indexOf("modulus: ") + 9;
+				if (start < 9) {
+					continue;
+				} else {
+					pubModulus = new BigInteger(part.substring(start));
+					break;
+				}
+			}
+		}
+		if (privModulus.equals(pubModulus)) {
```

Comment:
similar to my comment above, should we log a better error here if pubModulus is null?

Reply:
I corrected it in the latest commit. This case will fall through the error message that is output by the calling method.